### PR TITLE
[FLINK-12673][network] Introduce NetworkEnvironment.getUnreleasedPartitions instead of using getResultPartitionManager

### PR DESCRIPTION
**Pull request by azagrebin**
## What is the purpose of the change

At the moment NetworkEnvironment.getResultPartitionManager is used to query TaskExecutor whether all partitions are released. We should hide this direct access to ResultPartitionManager behind a Shuffle API method call: NetworkEnvironment.getUnreleasedPartitions.

## Brief change log

 - Refactor `ResultPartitionManager.areAllPartitionsReleased` to `getUnreleasedPartitions`
 - Add `NetworkEnvironment.getUnreleasedPartitions`

## Verifying this change

Trivial refactor covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by azagrebin** [comment]
@flinkbot attention @zhijiangW @zentol

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java [hunk 1]
```diff
@@ -122,14 +123,9 @@ void onConsumedPartition(ResultPartition partition) {
...
-	public boolean areAllPartitionsReleased() {
+	public Collection<ResultPartitionID> getUnreleasedPartitions() {
 		synchronized (registeredPartitions) {
-			for (ResultPartition partition : registeredPartitions.values()) {
-				if (!partition.isReleased()) {
-					return false;
-				}
-			}
-			return true;
+			return registeredPartitions.keySet();
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java]: The behavior here is different from before. We should confirm the criteria whether based on collection size or real partition state. It exists the scenario that the partition is already removed from the collection but the state is not released yet during`ResultPartitionManager#onConsumedPartition`. That means if we release the task executor immediately, it might lose chance for releasing the partition resource gracefully. WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 2]
```diff
@@ -201,6 +203,18 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
 		}
 	}
 
+	/**
+	 * Query unreleased partitions.
+	 *
+	 * @return collection of partitions which still occupy some resources locally on this task executor
+	 * and have not been released yet. The partition can be released either with {@link ResultPartitionWriter#fail(Throwable)}
+	 * or {@link ResultPartitionWriter#finish()} and then with {@link ResultPartitionWriter#close()} after fail or finish.
```
Comment (zhijiangW) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: It might bring some confusing here.
`ResultPartitionWriter` has two different states `FINISHED` and `RELEASED` separately. `ResultPartitionWriter#finish()` only indicates the whole data produced and enters `FINISHED` state, but the partition is not released yet. The release could only happen in three scenarios atm

- `ResultPartitionWriter#fail` when task is failed or canceled
- `NetworkEnvironment#releasePartitions` from JM RPC call
- `ResultPartitionManager#onConsumedPartition` when partition is consumed by consumer side

In addition, it might be better to put these comments above and make `@return` short.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java (hunk 1 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java]: Ok, the behaviour did not actually change but this problem was already before :) If we remove partition in `ResultPartitionManager#onConsumedPartition` and concurrently continue to release it, the old `areAllPartitionsReleased` will still skip its state check because it is simply not  in the `registeredPartitions` anymore.

I think we have to fix it here by moving release under the lock as it is already done in `ResultPartitionManager.releasePartition`.

In general, I think `registeredPartitions` should reflect the release state of them. At the end, memory occupied by partition in `registeredPartitions` is also a resource to release so being in `registeredPartitions` and unreleased should be always the case. If partition is released/unregistered, there is no point to keep in `registeredPartitions` either.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 2 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: In this comment I was actually thinking more about how to describe which user calls of shuffle API will lead to the release. `ResultPartitionWriter#finish()` triggers end of production and will result eventually internally in `ResultPartitionManager#onConsumedPartition`. But I think you are right we should mention that the actual release can happen later e.g. when the consumption is done. I will rewrite the comment a bit.

In general, `Close` could also potentially release some resources associated with partition, like memory now, we just do not associate it with release state of partition now because of current implementation details.

**Comment by azagrebin** [comment]
Thanks for the review @zhijiangW ! I addressed the comments.

As for returning a collection of partitions, it is a simple step to prepare for partition lifecycle management where we plan that the `getUnreleasedPartitions` will become part of TM/RM heartbeat where TM reports its currently being held partitions.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 3]
```diff
@@ -201,6 +203,28 @@ public void releasePartitions(Collection<ResultPartitionID> partitionIds) {
 		}
 	}
 
+	/**
+	 * Report unreleased partitions.
+	 *
+	 * <p>Partitions are released in the following cases:
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: I don't believe this javadoc belongs on this method. It contains information about implementation details of various methods, and thus should likely be placed into the class javadoc.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java [hunk 4]
```diff
@@ -172,6 +173,7 @@ private static void registerNetworkMetrics(MetricGroup metricGroup, NetworkBuffe
 	//  Properties
 	// --------------------------------------------------------------------------------------------
 
+	@VisibleForTesting
 	public ResultPartitionManager getResultPartitionManager() {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: can this be package private?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 3 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: This is mostly a prep step for introducing a shuffle service interface, it is not intended only for netty implementation. We can think where it belongs when the interface is introduced. I would keep it here now because it is mostly relevant here atm to understand what is unreleased.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 4 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: No, it is used by `StreamNetworkBenchmarkEnvironment` from other package atm. It is netty specific test performance benchmark.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java (hunk 1 above)
Comment (zhijiangW) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java]: Yes, the `partition#release` should also put under the synchronized in `ResultPartitionManager#onConsumedPartition`, then it could solve this issue.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java [hunk 5]
```diff
@@ -106,30 +107,23 @@ public void shutdown() {
...
-		}
-
-		// Release the partition if it was successfully removed
-		if (partition == previous) {
-			partition.release();
-
-			LOG.debug("Released {}.", partition);
+			final ResultPartition previous = registeredPartitions.remove(partition.getPartitionId());
+			// Release the partition if it was successfully removed
+			if (partition == previous) {
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionManager.java]: this change exceeds the scope of the original issue and should be put into a separate commit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java (hunk 3 above)
Comment (azagrebin) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java]: Let's remove this doc comment for now in favour of more comprehensive description given in #8608.

**Comment by azagrebin** [comment]
Thanks for the reviews @zhijiangW @zentol ! I have addressed the comments.
